src/backend/distributed/executor/directed_acylic_graph_execution.c

```diff
@@ -0,0 +1,218 @@
+	ListCell *taskCell = NULL;
+
+
+	foreach(taskCell, allTasks)
+	{
+		Task *task = (Task *) lfirst(taskCell);
+
+		if (IsAllDependencyCompleted(task, completedTasks) &&
+			!IsTaskAlreadyCompleted(task, completedTasks))
+		{
+			curTasks = lappend(curTasks, task);
+		}
+	}
+
+	return curTasks;
+}
+
+
+/*
+ * AddCompletedTasks adds the givens tasks to completedTasks HTAB.
+ */
+static void
+AddCompletedTasks(List *curCompletedTasks, HTAB *completedTasks)
+{
+	ListCell *taskCell = NULL;
+
+	bool found;
+
+	foreach(taskCell, curCompletedTasks)
+	{
+		Task *task = (Task *) lfirst(taskCell);
+		TaskHashKey taskKey = { task->jobId, task->taskId };
+		hash_search(completedTasks, &taskKey, HASH_ENTER, &found);
+	}
+}
+
+
+/*
+ * CreateTaskHashTable creates a HTAB with the necessary initialization.
+ */
+static HTAB *
+CreateTaskHashTable()
+{
+	uint32 hashFlags = (HASH_ELEM | HASH_FUNCTION | HASH_CONTEXT | HASH_COMPARE);
+	HASHCTL info = InitHashTableInfo();
+	return hash_create("citus task completed list (jobId, taskId)",
+					   64, &info, hashFlags);
+}
+
+
+/*
+ * IsTaskAlreadyCompleted returns true if the given task
+ * is found in the completedTasks HTAB.
+ */
+static bool
+IsTaskAlreadyCompleted(Task *task, HTAB *completedTasks)
+{
+	bool found;
+
+	TaskHashKey taskKey = { task->jobId, task->taskId };
+	hash_search(completedTasks, &taskKey, HASH_ENTER, &found);
+	return found;
+}
+
+
+/*
+ * IsAllDependencyCompleted return true if the given task's
+ * dependencies are completed.
+ */
+static bool
+IsAllDependencyCompleted(Task *targetTask, HTAB *completedTasks)
+{
+	ListCell *taskCell = NULL;
+	bool found = false;
+
+
+	foreach(taskCell, targetTask->dependentTaskList)
+	{
+		Task *task = (Task *) lfirst(taskCell);
+		TaskHashKey taskKey = { task->jobId, task->taskId };
+
+		hash_search(completedTasks, &taskKey, HASH_FIND, &found);
+		if (!found)
+		{
+			return false;
+		}
+	}
+	return true;
+}
+
+
+/*
+ * InitHashTableInfo returns hash table info, the hash table is
+ * configured to be created in the CurrentMemoryContext so that
+ * it will be cleaned when this memory context gets freed/reset.
+ */
+static HASHCTL
+InitHashTableInfo()
+{
+	HASHCTL info;
+
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(TaskHashKey);
+	info.entrysize = sizeof(TaskHashEntry);
+	info.hash = TaskHash;
+	info.match = TaskHashCompare;
+	info.hcxt = CurrentMemoryContext;
+
+	return info;
+}
+
+
+static uint32
+TaskHash(const void *key, Size keysize)
+{
+	TaskHashKey *taskKey = (TaskHashKey *) key;
+	uint32 hash = 0;
+
+	hash = hash_combine(hash, hash_any((unsigned char *) &taskKey->jobId,
+									   sizeof(int64)));
+	hash = hash_combine(hash, hash_uint32(taskKey->taskId));
+
+	return hash;
+}
+
+
+static int
+TaskHashCompare(const void *key1, const void *key2, Size keysize)
+{
+	TaskHashKey *taskKey1 = (TaskHashKey *) key1;
+	TaskHashKey *taskKey2 = (TaskHashKey *) key2;
+	return taskKey1->jobId != taskKey2->jobId || taskKey1->taskId != taskKey2->taskId;
```

Comment:
This does not seem like a correct compare function. It's supposed to return -1, 0 or 1.